### PR TITLE
Adding flatbuffers and migrating flatcc to flatbuffers for plasma

### PR DESCRIPTION
**Pull request by atumanov**

### CODE REVIEW

File: build.sh [hunk 1]
```diff
@@ -25,7 +25,7 @@ bash "$ROOT_DIR/src/numbuf/thirdparty/build_thirdparty.sh"
 
 # Now build everything.
 pushd "$ROOT_DIR/python/core"
-  cmake ../..
+  cmake -DCMAKE_BUILD_TYPE=Release  ../..
```
Comment (robertnishihara) [inline build.sh]: will this cause all of the `DCHECK`'s to stop working?

File: src/common/CMakeLists.txt [hunk 2]
```diff
@@ -34,7 +34,7 @@ target_link_libraries(common "${CMAKE_CURRENT_LIST_DIR}/thirdparty/hiredis/libhi
 
 function(define_test test_name library)
   add_executable(${test_name} test/${test_name}.c ${ARGN})
-  add_dependencies(${test_name} hiredis flatcc)
+  add_dependencies(${test_name} hiredis flatcc flatbuffers_ep)
   target_link_libraries(${test_name} common ${FLATBUFFERS_STATIC_LIB} ${library})
   target_compile_options(${test_name} PUBLIC "-DPLASMA_TEST -DPHOTON_TEST -DCOMMON_TEST -DRAY_COMMON_LOG_LEVEL=4 -DRAY_TIMEOUT=50")
 endfunction()
```
Comment (robertnishihara) [inline src/common/CMakeLists.txt]: Removing `src/common/build/.gitkeep` (not this file, but the next one in the PR) was unintentional, right?

Comment (robertnishihara) [inline src/common/CMakeLists.txt]: Same with `src/global_scheduler/build/.gitkeep` and `src/photon/build/.gitkeep`

File: src/plasma/CMakeLists.txt [hunk 3]
```diff
@@ -51,40 +66,53 @@ if(APPLE)
                      POST_BUILD COMMAND ${CMAKE_INSTALL_NAME_TOOL} -change ${PYTHON_SHARED_LIBRARY} ${PYTHON_LIBRARIES} libplasma.so)
 endif(APPLE)
 
-target_link_libraries(plasma ${COMMON_LIB} ${PYTHON_LIBRARIES} ${FLATBUFFERS_STATIC_LIB})
+if(APPLE)
```
Comment (robertnishihara) [inline src/plasma/CMakeLists.txt]: The statements inside if the 

`if(APPLE)`
`else(APPLE)`
`endif(APPLE)`

should be indented two spaces

File: src/plasma/plasma_protocol.c [hunk 4]
```diff
@@ -76,7 +76,10 @@ int finalize_buffer_and_send(flatcc_builder_t *B, int fd, int message_type) {
   void *buff = flatcc_builder_finalize_buffer(B, &size);
   int r = write_message(fd, message_type, size, buff);
   free(buff);
-  flatcc_builder_reset(B);
+  if (!(message_type == MessageType_PlasmaCreateRequest || message_type == MessageType_PlasmaSealRequest)) {
```
Comment (robertnishihara) [inline src/plasma/plasma_protocol.c]: Is this still relevant?

File: src/plasma/plasma_protocol.cc [hunk 5]
```diff
@@ -0,0 +1,541 @@
...
+
+void plasma_read_CreateRequest(uint8_t *data,
+                               object_id *object_id,
+                               int64_t *data_size,
+                               int64_t *metadata_size) {
+  CHECK(data);
+  auto message = flatbuffers::GetRoot<PlasmaCreateRequest>(data);
+  *data_size = message->data_size();
+  *metadata_size = message->metadata_size();
+  CHECK(message->object_id()->size() == UNIQUE_ID_SIZE);
```
Comment (robertnishihara) [inline src/plasma/plasma_protocol.cc]: We probably shouldn't use the `UNIQUE_ID_SIZE` macro, but instead use e.g., `sizeof(*object_id)`.

File: src/plasma/plasma_protocol.cc [hunk 6]
```diff
@@ -0,0 +1,541 @@
...
+  fbb.Finish(message);
+  return write_message(sock, MessageType_PlasmaDataRequest, fbb.GetSize(), fbb.GetBufferPointer());
+}
+
+void plasma_read_DataRequest(uint8_t *data,
+                             object_id *object_id,
+                             char **address,
+                             int *port) {
+  DCHECK(data);
+  /*  */
```
Comment (robertnishihara) [inline src/plasma/plasma_protocol.cc]: `/* */` is unused

File: src/common/CMakeLists.txt (hunk 2 above)
Comment (pcmoritz) [inline src/common/CMakeLists.txt]: These .gitkeeps are not used any more. Not sure it should be part of this PR but they should be removed...

Comment (robertnishihara) [inline src/common/CMakeLists.txt]: Oh I see.

File: build.sh (hunk 1 above)
Comment (pcmoritz) [inline build.sh]: No

**Review by stephanie-wang — APPROVED** [review]

**Comment by pcmoritz** [comment]
Oh wow, thanks for catching that :1st_place_medal:

**Pull request merged**